### PR TITLE
feat(examples): Add a streamlit mnist example

### PR DESCRIPTION
**Pull request by gaocegege**
Peek 2022-08-18 15-01

Signed-off-by: Ce Gao <cegao@tensorchord.ai>

### CODE REVIEW

**Review by aseaday — APPROVED** [review]

**Comment by aseaday** [comment]
曾经沧海难为水，又见识别手写数

File: examples/streamlit-mnist/build.envd [hunk 1]
```diff
@@ -0,0 +1,34 @@
+def build():
+    base(os="ubuntu20.04", language="python3")
+    install.vscode_extensions([
+        "ms-python.python",
+    ])
+
+    # Install packages.
+    config.pip_index(url = "https://pypi.tuna.tsinghua.edu.cn/simple")
```
Comment (kemingy) [inline examples/streamlit-mnist/build.envd:8]: Should we comment this line?

File: examples/streamlit-mnist/build.envd [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        "streamlit",
+        "streamlit_drawable_canvas",
+    ])
+    runtime.expose(envd_port=port, host_port=port, service="streamlit")
+
+def configure_mnist():
+    install.system_packages([
+        "libgl1",
+    ])
+    install.python_packages([
```
Comment (kemingy) [inline examples/streamlit-mnist/build.envd:29]: Can we reuse the `requirements.txt` file?

File: examples/streamlit-mnist/train.ipynb [hunk 3]
```diff
@@ -0,0 +1,343 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "ename": "ModuleNotFoundError",
+     "evalue": "No module named 'matplotlib'",
```
Comment (kemingy) [inline examples/streamlit-mnist/train.ipynb:10]: Can clean this notebook.

File: examples/streamlit-mnist/app.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from tensorflow.keras.models import load_model
+import streamlit as st
+from streamlit_drawable_canvas import st_canvas
+
+MODEL_DIR = os.path.join(os.path.dirname(__file__), 'model')
+if not os.path.isdir(MODEL_DIR):
+    os.system('runipy train.ipynb')
+
+model = load_model('model')
+# st.markdown('<style>body{color: White; background-color: DarkSlateGrey}</style>', unsafe_allow_html=True)
```
Comment (kemingy) [inline examples/streamlit-mnist/app.py:13]: Can delete the test code

File: examples/streamlit-mnist/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+[streamlit](https://www.streamlit.io/) is an open-source app framework, which is the easiest way for data scientists and machine learning engineers to create beautiful, performant apps. All in pure Python, no longer fiddling with javascript.
+
+This demo contains two parts: training a simple digit recognition model using mnist dataset and a webapp to live demo that model.
+ 
+## Running demo
+
+1. First install all the dependencies
+
+    ```
+    pip install -r requirements.txt
```
Comment (kemingy) [inline examples/streamlit-mnist/README.md:16]: `envd up` should install everything if we use the `requirements.txt` in `build.evnd`

File: examples/streamlit-mnist/build.envd (hunk 1 above)
Comment (gaocegege) [reply examples/streamlit-mnist/build.envd:8]: Yep, I think so.

File: examples/streamlit-mnist/build.envd (hunk 2 above)
Comment (gaocegege) [reply examples/streamlit-mnist/build.envd:29]: I think we can remove requirements.txt, WDYT

File: examples/streamlit-mnist/train.ipynb (hunk 3 above)
Comment (gaocegege) [reply examples/streamlit-mnist/train.ipynb:10]: OK.

File: examples/streamlit-mnist/app.py (hunk 4 above)
Comment (gaocegege) [reply examples/streamlit-mnist/app.py:13]: OK.

File: examples/streamlit-mnist/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+numpy
+matplotlib
+tensorflow
+runipy
+opencv-python
+streamlit==0.64.0
+streamlit_drawable_canvas==0.4.0
```
Comment (kemingy) [inline examples/streamlit-mnist/requirements.txt:7]: These two are outdated.

File: pkg/docker/docker.go [hunk 7]
```diff
@@ -557,7 +557,7 @@ func (c generalClient) StartEnvd(ctx context.Context, tag, name, buildContext st
 		// Hack to check if the port is already allocated.
 		if strings.Contains(errCause.Error(), "port is already allocated") {
 			logrus.Debugf("failed to allocate the port: %s", err)
-			return "", "", errors.New("jupyter port is already allocated in the host")
+			return "", "", errors.Newf("port is already allocated in the host")
```
Comment (kemingy) [inline pkg/docker/docker.go:560]:
```suggestion
			return "", "", errors.New("port is already allocated in the host")
```

File: examples/streamlit-mnist/build.envd (hunk 2 above)
Comment (kemingy) [reply examples/streamlit-mnist/build.envd:33]: Yep

**Comment by gaocegege** [comment]
PTAL

File: examples/streamlit-mnist/app.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import os
+import numpy as np
+import cv2
+from tensorflow.keras.models import load_model
+import streamlit as st
+from streamlit_drawable_canvas import st_canvas
+
+MODEL_DIR = os.path.join(os.path.dirname(__file__), 'model')
+if not os.path.isdir(MODEL_DIR):
+    os.system('runipy train.ipynb')
```
Comment (kemingy) [inline examples/streamlit-mnist/app.py:10]: Can delete this part.

Comment (kemingy) [reply examples/streamlit-mnist/app.py:10]: Shall we add a license for this file?

Comment (gaocegege) [reply examples/streamlit-mnist/app.py:10]: The original repo does not contain a license, thus I am not sure what license we should add.

**Review by kemingy — APPROVED** [review]

**Pull request merged**